File: frontend/src/TerraForge/lib/shapes.js
// Per-preset shape post-processing. Applied AFTER noise generation to make each preset
// visually distinct (instead of just being a noise param tweak).
// All shape functions take heightmap in place and return it. h is Float32Array 0..1.

function smoothstep(e0, e1, x) {
  const t = Math.max(0, Math.min(1, (x - e0) / (e1 - e0 || 1e-6)));
  return t * t * (3 - 2 * t);
}

function normalize(h) {
  let mn = Infinity, mx = -Infinity;
  for (let i = 0; i < h.length; i++) { if (h[i] < mn) mn = h[i]; if (h[i] > mx) mx = h[i]; }
  const r = mx - mn || 1;
  for (let i = 0; i < h.length; i++) h[i] = (h[i] - mn) / r;
  return h;
}

// Radial cone mask: 1 at center, 0 at edges. Power controls steepness.
function radialMask(size, power = 2.0, innerRadius = 0.0) {
  const mask = new Float32Array(size * size);
  const cx = (size - 1) / 2;
  const cy = (size - 1) / 2;
  const maxR = Math.sqrt(cx * cx + cy * cy);
  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      const d = Math.hypot(x - cx, y - cy) / maxR;
      let v = 1 - smoothstep(innerRadius, 1.0, d);
      v = Math.pow(v, power);
      mask[y * size + x] = v;
    }
  }
  return mask;
}

// Identity (no post-process)
export function shapeIdentity(h) { return h; }

// Alpine: sharpen peaks
export function shapeAlpine(h) {
  for (let i = 0; i < h.length; i++) {
    h[i] = Math.pow(h[i], 0.85);
  }
  return normalize(h);
}

// Rolling Hills: soften, push to mid-range
export function shapeRolling(h) {
  for (let i = 0; i < h.length; i++) {
    h[i] = Math.pow(h[i], 1.1) * 0.7 + 0.15;
  }
  return normalize(h);
}

// Desert Mesa: flat-topped plateaus (clamp + step pattern)
export function shapeMesa(h) {
  for (let i = 0; i < h.length; i++) {
    let v = h[i];
    // Quantize into 4 plateau levels with smooth edges
    const levels = 4;
    const lv = v * levels;
    const base = Math.floor(lv);
    const frac = lv - base;
    const edge = smoothstep(0.85, 1.0, frac) * 0.15;
    v = (base + edge) / levels + frac * 0.05;
    h[i] = v;
  }
  return normalize(h);
}

// Volcanic Island: radial cone + crater dip in center + lower to water at edges
export function shapeVolcanic(h, size) {
  const mask = radialMask(size, 1.8, 0.1);
  const cx = (size - 1) / 2;
  const cy = (size - 1) / 2;
  const maxR = Math.sqrt(cx * cx + cy * cy);
  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      const i = y * size + x;
      const d = Math.hypot(x - cx, y - cy) / maxR;
      let v = h[i];
      // multiply by cone mask
      v = v * 0.4 + mask[i] * 0.7;
      // crater dip near center
      const craterT = smoothstep(0, 0.08, d);
      const craterDepth = 0.25 * (1 - craterT);
      v -= craterDepth;
      h[i] = Math.max(0, v);
    }
  }
  return normalize(h);
}

// Archipelago: radial falloff to water + bias many small islands
export function shapeArchipelago(h, size) {
  const cx = (size - 1) / 2;
  const cy = (size - 1) / 2;
  const maxR = Math.sqrt(cx * cx + cy * cy);
  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      const i = y * size + x;
      const d = Math.hypot(x - cx, y - cy) / maxR;
      // pull edges underwater
      const edgeFalloff = 1 - smoothstep(0.55, 1.0, d);
      // emphasize highs (islands) and depress lows (water)
      let v = h[i];
      v = Math.pow(v, 1.4); // steepen
      v = v * edgeFalloff * 0.85 + 0.05;
      h[i] = v;
    }
  }
  return normalize(h);
}

// Canyon: take inverted ridged noise and gouge linear cuts across.
// Apply strong vertical/horizontal carving inspired by river valleys.
export function shapeCanyon(h, size) {
  // Carving pattern: a wavy "river" channel along x using a sine-perturbed centerline.
  // Wherever a pixel is close to the channel, we lower the height.
  for (let y = 0; y < size; y++) {
    for (let x = 0; x < size; x++) {
      const i = y * size + x;
      // 2 winding channels
      const ch1 = size * 0.4 + Math.sin(x * 0.025) * size * 0.12 + Math.sin(x * 0.07) * size * 0.04;
      const ch2 = size * 0.7 + Math.sin(x * 0.018 + 1.3) * size * 0.10;
      const d1 = Math.abs(y - ch1);
      const d2 = Math.abs(y - ch2);
      const d = Math.min(d1, d2);
      const carveRadius = size * 0.03;
      const carve = 1 - smoothstep(carveRadius * 0.4, carveRadius * 2.2, d);
      let v = h[i];
      // High elevation overall, then carve deep narrow valleys
      v = Math.pow(v, 0.9) * 0.7 + 0.25;
      v -= carve * 0.55;
      h[i] = Math.max(0, v);
    }
  }
  return normalize(h);
}

// High Plateau: lift mid-range, flatten top
export function shapePlateau(h) {
  for (let i = 0; i < h.length; i++) {
    let v = h[i];
    // S-curve: lows stay low, mids lift fast, highs plateau
    v = smoothstep(0.15, 0.55, v) * 0.7 + v * 0.3;
    // flatten near top
    if (v > 0.7) v = 0.7 + (v - 0.7) * 0.3;
    h[i] = v;
  }
  return normalize(h);
}

// Dispatch by preset id
export function applyPresetShape(presetId, h, size) {
  switch (presetId) {
    case "alpine": return shapeAlpine(h);
    case "rolling": return shapeRolling(h);
    case "desert": return shapeMesa(h);
    case "volcanic": return shapeVolcanic(h, size);
    case "archipelago": return shapeArchipelago(h, size);
    case "canyon": return shapeCanyon(h, size);
    case "plateau": return shapePlateau(h);
    default: return h;
  }
}

File: TerraForgeRust/core/src/shapes.rs
//! Preset shape post-processing. Port of `lib/shapes.js`.

use crate::{normalize, smoothstep};

fn radial_mask(size: u32, power: f32, inner_radius: f32) -> Vec<f32> {
    let n = size as usize;
    let mut mask = vec![0f32; n * n];
    let cx = (n as f32 - 1.0) * 0.5;
    let cy = (n as f32 - 1.0) * 0.5;
    let max_r = (cx * cx + cy * cy).sqrt();
    for y in 0..n {
        for x in 0..n {
            let d = ((x as f32 - cx).powi(2) + (y as f32 - cy).powi(2)).sqrt() / max_r;
            let v = (1.0 - smoothstep(inner_radius, 1.0, d)).powf(power);
            mask[y * n + x] = v;
        }
    }
    mask
}

pub fn shape_identity(_h: &mut [f32]) {}

pub fn shape_alpine(h: &mut [f32]) {
    for v in h.iter_mut() { *v = (*v).powf(0.85); }
    normalize(h);
}

pub fn shape_rolling(h: &mut [f32]) {
    for v in h.iter_mut() { *v = (*v).powf(1.1) * 0.7 + 0.15; }
    normalize(h);
}

pub fn shape_mesa(h: &mut [f32]) {
    for v in h.iter_mut() {
        let levels = 4.0f32;
        let lv = *v * levels;
        let base = lv.floor();
        let frac = lv - base;
        let edge = smoothstep(0.85, 1.0, frac) * 0.15;
        *v = (base + edge) / levels + frac * 0.05;
    }
    normalize(h);
}

pub fn shape_volcanic(h: &mut [f32], size: u32) {
    let n = size as usize;
    let mask = radial_mask(size, 1.8, 0.1);
    let cx = (n as f32 - 1.0) * 0.5;
    let cy = (n as f32 - 1.0) * 0.5;
    let max_r = (cx * cx + cy * cy).sqrt();
    for y in 0..n {
        for x in 0..n {
            let i = y * n + x;
            let d = ((x as f32 - cx).powi(2) + (y as f32 - cy).powi(2)).sqrt() / max_r;
            let mut v = h[i] * 0.4 + mask[i] * 0.7;
            let crater_t = smoothstep(0.0, 0.08, d);
            v -= 0.25 * (1.0 - crater_t);
            h[i] = v.max(0.0);
        }
    }
    normalize(h);
}

pub fn shape_archipelago(h: &mut [f32], size: u32) {
    let n = size as usize;
    let cx = (n as f32 - 1.0) * 0.5;
    let cy = (n as f32 - 1.0) * 0.5;
    let max_r = (cx * cx + cy * cy).sqrt();
    for y in 0..n {
        for x in 0..n {
            let i = y * n + x;
            let d = ((x as f32 - cx).powi(2) + (y as f32 - cy).powi(2)).sqrt() / max_r;
            let edge_falloff = 1.0 - smoothstep(0.55, 1.0, d);
            let mut v = h[i].powf(1.4);
            v = v * edge_falloff * 0.85 + 0.05;
            h[i] = v;
        }
    }
    normalize(h);
}

pub fn shape_canyon(h: &mut [f32], size: u32) {
    let n = size as usize;
    for y in 0..n {
        for x in 0..n {
            let i = y * n + x;
            let ch1 = n as f32 * 0.4 + ((x as f32) * 0.025).sin() * n as f32 * 0.12 + ((x as f32) * 0.07).sin() * n as f32 * 0.04;
            let ch2 = n as f32 * 0.7 + ((x as f32) * 0.018 + 1.3).sin() * n as f32 * 0.10;
            let d1 = (y as f32 - ch1).abs();
            let d2 = (y as f32 - ch2).abs();
            let d = d1.min(d2);
            let carve_r = n as f32 * 0.03;
            let carve = 1.0 - smoothstep(carve_r * 0.4, carve_r * 2.2, d);
            let mut v = h[i].powf(0.9) * 0.7 + 0.25;
            v -= carve * 0.55;
            h[i] = v.max(0.0);
        }
    }
    normalize(h);
}

pub fn shape_plateau(h: &mut [f32]) {
    for v in h.iter_mut() {
        let mut x = *v;
        x = smoothstep(0.15, 0.55, x) * 0.7 + x * 0.3;
        if x > 0.7 { x = 0.7 + (x - 0.7) * 0.3; }
        *v = x;
    }
    normalize(h);
}

/// Apply the shape associated with a preset id. Unknown ids are no-ops.
pub fn apply_preset_shape(preset_id: &str, h: &mut [f32], size: u32) {
    match preset_id {
        "alpine" => shape_alpine(h),
        "rolling" => shape_rolling(h),
        "desert" => shape_mesa(h),
        "volcanic" => shape_volcanic(h, size),
        "archipelago" => shape_archipelago(h, size),
        "canyon" => shape_canyon(h, size),
        "plateau" => shape_plateau(h),
        _ => shape_identity(h),
    }
}
